guide/src/memory.md

```diff
@@ -0,0 +1,185 @@
+`GILPool` which owns the memory pointed to by the reference.  In the example
+above, the lifetime of the reference `hello` is bound to the `GILPool`.  When
+the `with_gil()` closure ends or the `GILGuard` from `acquire_gil()` is dropped,
+the `GILPool` is also dropped and the Python reference counts of the variables
+it owns are decreased, releasing them to the Python garbage collector.  Most
+of the time we don't have to think about this, but consider the following:
+
+```rust
+Python::with_gil(|py| -> PyResult<()> {
+    for _ in 0..10 {
+        let hello: &PyString = py.eval("\"Hello World!\"", None, None)?.extract()?;
+        println!("Python says: {}", hello);
+    }
+    // There are 10 copies of `hello` on Python's heap here.
+    Ok(())
+})?;
+```
+
+We might assume that the `hello` variable's memory is freed at the end of each
+loop iteration, but in fact we create 10 copies of `hello` on Python's heap.
+This may seem surprising at first, but it is completely consistent with Rust's
+memory model.  The `hello` variable is dropped at the end of each loop, but it
+is only a reference to the memory owned by the `GILPool`, and its lifetime is
+bound to the `GILPool`, not the for loop.  The `GILPool` isn't dropped until
+the end of the `with_gil()` closure, at which point the 10 copies of `hello`
+are finally released to the Python garbage collector.
+
+In general we don't want unbounded memory growth during loops!  One workaround
+is to acquire and release the GIL with each iteration of the loop.
+
+```rust
+for _ in 0..10 {
+    Python::with_gil(|py| -> PyResult<()> {
+        let hello: &PyString = py.eval("\"Hello World!\"", None, None)?.extract()?;
+        println!("Python says: {}", hello);
+        Ok(())
+    })?; // only one copy of `hello` at a time
+}
+```
+
+It might not be practical or performant to acquire and release the GIL so many
+times.  Another workaround is to work with the `GILPool` object directly, but
+this is unsafe.
+
+```rust
+Python::with_gil(|py| -> PyResult<()> {
+    for _ in 0..10 {
+        let pool = unsafe { py.new_pool() };
+        let py = pool.python();
+        let hello: &PyString = py.eval("\"Hello World!\"", None, None)?.extract()?;
+        println!("Python says: {}", hello);
+    }
+    Ok(())
+})?;
+```
+
+The unsafe method `Python::new_pool` allows you to create a nested `GILPool`
+from which you can retrieve a new `py: Python` GIL token.  Variables created
+with this new GIL token are bound to the nested `GILPool` and will be released
+when the nested `GILPool` is dropped.  Here, the nested `GILPool` is dropped
+at the end of each loop iteration, before the `with_gil()` closure ends.
+
+When doing this, you must be very careful to ensure that once the `GILPool` is
+dropped you do not retain access any owned references created after the
+`GILPool` was created.  Read the documentation for `Python::new_pool()` for more
```

Comment:
It would be great to add a link to the docs on `Python::new_pool()` here.

Reply:
Link added.